File: backend/app/schemas/issue.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, validator

from app.models.issue import IssuePriority, IssueStatus, IssueType


def _ensure_tags(value) -> List[str]:
    if isinstance(value, list):
        return [tag for tag in value if tag]
    if isinstance(value, str):
        return [tag.strip() for tag in value.split(",") if tag.strip()]
    return []


class IssueBase(BaseModel):
    title: str
    description: str
    type: IssueType = IssueType.BUG
    priority: IssuePriority = IssuePriority.MEDIUM
    status: IssueStatus = IssueStatus.OPEN
    tags: List[str] = Field(default_factory=list)
    location: Optional[str] = None

    _normalize_tags = validator("tags", allow_reuse=True, pre=True)(_ensure_tags)


class IssueCreate(IssueBase):
    repository_id: int
    priority: IssuePriority = IssuePriority.MEDIUM
    status: IssueStatus = IssueStatus.OPEN
    estimated_cost: Optional[int] = None
    budget_category: Optional[str] = None


class IssueUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    type: Optional[IssueType] = None
    priority: Optional[IssuePriority] = None
    status: Optional[IssueStatus] = None
    tags: Optional[List[str]] = None
    location: Optional[str] = None
    estimated_cost: Optional[int] = None
    budget_category: Optional[str] = None


class Issue(IssueBase):
    id: int
    number: int
    slug: str
    repository_id: int
    author_id: int
    assigned_to_id: Optional[int] = None
    comments_count: int
    reactions_count: int
    created_at: datetime
    updated_at: datetime
    closed_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/issue.ts
export type IssueStatus = 'open' | 'in_progress' | 'resolved' | 'closed' | 'duplicate'
export type IssuePriority = 'low' | 'medium' | 'high' | 'urgent'
export type IssueType = 'bug' | 'feature' | 'improvement' | 'policy' | 'infrastructure' | 'service'

export interface Issue {
  id: number
  number: number
  slug: string
  repository_id: number
  author_id: number
  assigned_to_id?: number | null
  title: string
  description: string
  type: IssueType
  status: IssueStatus
  priority: IssuePriority
  tags: string[]
  location?: string | null
  estimated_cost?: number | null
  budget_category?: string | null
  comments_count: number
  reactions_count: number
  created_at: string
  updated_at: string
  closed_at?: string | null
}

export interface IssueComment {
  id: number
  issue_id: number
  author_id: number
  content: string
  created_at: string
  updated_at: string
}
